native/cocos/renderer/gfx-base/GFXDeviceObject.h

```diff
@@ -0,0 +1,64 @@
+/****************************************************************************
+Copyright (c) 2020-2022 Xiamen Yaji Software Co., Ltd.
+http://www.cocos.com
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated engine source code (the "Software"), a limited,
+worldwide, royalty-free, non-assignable, revocable and non-exclusive license
+to use Cocos Creator solely to develop games on your target platforms. You shall
+not use Cocos Creator software for developing other software or tools that's
+used for developing games. You are not granted to publish, distribute,
+sublicense, and/or sell copies of Cocos Creator.
+The software or tools in this License Agreement are licensed, not sold.
+Xiamen Yaji Software Co., Ltd. reserves all rights not expressly granted to you.
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+****************************************************************************/
+
+#pragma once
+
+#include "base/Ptr.h"
+#include "base/Macros.h"
+
+namespace cc {
+namespace gfx {
+
+class GFXDeviceObject {
+public:
+    virtual ~GFXDeviceObject() = default;
+
+    void addRef() const {
+        ++_referenceCount;
+    }
+
+    void release() const {
+        CC_ASSERT(_referenceCount > 0);
+        --_referenceCount;
+
+        if (_referenceCount == 0) {
+            const_cast<GFXDeviceObject*>(this)->shutdown();
+            delete this;
+        }
+    }
+
+    unsigned int getRefCount() const {
+        return _referenceCount;
+    }
+
+protected:
+    virtual void shutdown() {}
+
+    GFXDeviceObject() = default;
+
+    mutable unsigned int _referenceCount{0};
```

Comment:
Mutable member has to be thread-safe. Consider using std::atomic_int32_t. I think signed int is better unsigned int, because it is easier to check addRef/release mismatch.

Reply:
Here I use mutable only for constness.

Reply:
Most of the GPUObjects are defiend with member variables. Do we need to use ```ConstPtr`` to restrict the modification of member variables?